Upgrade the Escrow to be feature rich (#54)
FuelLabs/sway-applications: escrow/SPECIFICATION.md

Table of Content
- [Actions that users are able to perform](#actions-that-users-are-able-to-perform)
  - [Arbiter](#arbiter)
    - [`resolve_dispute()`](#resolve_dispute)
  - [Buyer](#buyer)
    - [`accept_arbiter()`](#accept_arbiter)
    - [`deposit()`](#deposit)
    - [`dispute()`](#dispute)
    - [`transfer_to_seller()`](#transfer_to_seller)
  - [Seller](#seller)
    - [`create_escrow()`](#create_escrow)
    - [`propose_arbiter()`](#propose_arbiter)
    - [`return_deposit()`](#return_deposit)
    - [`take_payment()`](#take_payment)

# Overview

This document provides an overview of the application.

It outline the use cases, i.e. desirable functionality, in addition to requirements for the smart contract and the user interface.

# Use Cases

This section contains general information about the functionality of the application and thus does not touch upon any technical aspects.

If you are interested in a functional overview then this is the section for you.

## Actions that users are able to perform

This sub-section details what a user is able to do e.g. click a button and "x, y, z" happens.

### Arbiter

#### `resolve_dispute()`

> 1. When an escrow is in dispute the arbiter of the escrow can perform a resolution in either direction of the buyer or seller

Reviewer: ```suggestion
1. When an escrow is in dispute the arbiter of the escrow can perform a resolution in either the direction of the buyer or seller
```

Author: This is incorrect

Reviewer: I plugged both options into an online grammar tool and they both work.  I guess it's just a matter of opinion.

2. A resolution consists of
   1. Selecting either the buyer or seller to whom the deposited asset from the buyer will be sent to
   2. Selecting an amount to take as payment from the collateral provided by the seller

### Buyer

#### `accept_arbiter()`

1. The buyer is able to accept a proposition by the seller to change the arbiter or change the fee for the arbiter

> **NOTE** They are not required to do so however there may be instances where changing the arbiter / fee is favourable

#### `deposit()`

> 1. The buyer must deposit into the escrow one asset from the options provided by the seller

Reviewer: ```suggestion
1. The buyer must deposit into the escrow one asset from the list of approved assets provided by the seller
```

Author: This implies the seller has done something to approve an asset. It doesn't really make sense especially in the context of the buyer actually having a function to approve

> 2. The escrow accepts a 1 time deposit therefore the buyer cannot go back and deposit a different asset

Reviewer: ```suggestion
2. The escrow accepts a 1 time deposit.  Therefore the buyer cannot go back and deposit a different asset
```

Author: This is also incorrect

Reviewer: 1. hmm I'm not sure this is incorrect.  I think the "therefore" needs some sort of punctuation around it since you are using it to connect two [independent clauses](https://www.wikihow.com/Use-Therefore-in-a-Sentence), but also grammar is not my strongest lol.

Reviewer: Also I think the original sentence works well enough if you don't feel like changin.

Author: Same here. Looking at that link I don't know why they end the sentence and start with "Therefore". I'm pretty sure that is incorrect or rather it can be done but that's not the correct way to use it. Afaik it's meant to connect and the common use may / must involve a comma at times

Reviewer: Feel free to add punctuation or don't.  I think readers will be able to understand the original.

Reviewer 2: Deep into bike-shedding territory here, but technically, it's either:

> "The escrow accepts a one-time deposit.  Therefore, the buyer ... " 

or

> "The escrow accepts a one-time deposit; therefore, the buyer ... "

Author: hm, I thought something like the second one was the correct way to do it while the first is incorrect. Guess I was never taught the first way

> **NOTE** Once a buyer deposits they are unable to get the asset out by themselves. Either the arbiter must resolve in their favour or the seller must return the deposit. This is a safety mechanism.

#### `dispute()`

1. The buyer is able to `dispute()` the escrow which prevents the seller from taking their deposit after the deadline

> 2. Disputing means that there are 3 ways to move the deposit from the buyer out of the escrow

Reviewer: ```suggestion
2. Disputing means that the contract now allows 3 ways to move the deposit from the buyer out of the escrow
```

Author: Technically the contract allowed 3 ways before & 3 ways after the dispute but now one of the options has changed so "now" doesn't really make sense here

   1. The buyer transfers to seller
   2. The seller returns the deposit to the buyer

>    3. The arbiter may come in and resolve the dispute in favour of the buyer or seller

Reviewer: It may be useful to a reader to include the functions that perform the operations mentions (`transfer_to_seller`) etc

Author: > to include the functions that perform the operations mentions

I do not understand.

I am planning on adding diagrams which show a visual contract flow of all the use cases

Reviewer: I mean it may be useful to mention the functions that perform the 3 different ways to move the deposit from the buyer out of the escrow.  `transfer_to_seller`, `resolve_dispute`, and `return_deposit`.  Although if you add diagrams in the future this may not be a useful addition.

#### `transfer_to_seller()`

1. When the buyer is satisfied with the exchange then they are able to transfer the funds to the seller

> **NOTE** They are not required to do so however being a good samaritan is nice. The seller can take the payment later themselves if the escrow is not in dispute

### Seller

#### `create_escrow()`

> 1. Any user that creates an escrow is automatically considered as the seller

Reviewer: ```suggestion
1. Any user that creates an escrow is automatically considered the seller
```

> 2. Creating an escrow requires collateral as payment for a possible arbitration

Reviewer: ```suggestion
2. Creating an escrow requires depositing collateral as payment for a possible arbitration
```

> 3. When creating an escrow the seller can choose any assets that they are willing to take as payment from the buyer

Reviewer: ```suggestion
3. When creating an escrow the seller provides a list of assets that they are willing to accept as payment from the buyer
```

#### `propose_arbiter()`

1. The seller may propose a change to the arbiter or arbiter fee
   1. This requires a new deposit as collateral
2. This can be done an unlimited number of times
   1. If this is done more than once in a row before the buyer accepts then the previous collateral will automatically be returned to the seller
3. If the buyer accepts then the previous collateral will be returned to the seller

> 4. If the buyer does not accept then upon completion of the escrow the funds will be returned to the seller

Reviewer: ```suggestion
4. If the buyer does not accept then upon completion of the escrow the unused collateral will be returned to the seller
```
